File: app/src/main/java/com/botob/ulisten2/services/MediaNotificationListenerService.java
package com.botob.ulisten2.services;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.speech.tts.Voice;
import android.support.v4.content.LocalBroadcastManager;
import android.text.TextUtils;
import android.util.Log;

import com.botob.ulisten2.MainActivity;
import com.botob.ulisten2.media.Media;
import com.botob.ulisten2.media.MediaApp;
import com.botob.ulisten2.media.MediaFactory;
import com.botob.ulisten2.notification.Extractor;
import com.botob.ulisten2.notification.NotificationData;
import com.botob.ulisten2.preferences.SettingsManager;

import java.util.Arrays;
import java.util.Locale;

public class MediaNotificationListenerService extends NotificationListenerService
        implements SharedPreferences.OnSharedPreferenceChangeListener {
    private static final String TAG = MediaNotificationListenerService.class.getSimpleName();

    private AudioManager mAudioManager;

    private TextToSpeech mTts;

    private Handler mHandler;

    private Media mCurrentMedia;

    private SettingsManager mSettingsManager;

    @Override
    public void onCreate() {
        Log.i(TAG, "onCreate");
        super.onCreate();
        // Instantiate the settings manager.
        mSettingsManager = new SettingsManager(getApplicationContext());

        // Register this service to listen to preference changes.
        mSettingsManager.getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(TAG, "onStartCommand");
        super.onStartCommand(intent, flags, startId);
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "onDestroy");
        if (mSettingsManager != null) {
            // Unregister this service from listening to preference changes.
            mSettingsManager.getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        Log.i(TAG, "onBind");
        resume();
        if (SERVICE_INTERFACE.equals(intent.getAction())) {
            return super.onBind(intent);
        } else {
            return new LocalBinder();
        }
    }

    @Override
    public void onRebind(Intent intent) {
        super.onRebind(intent);
        Log.i(TAG, "onRebind");
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Log.i(TAG, "onUnbind");
        return true;
    }

    public void resume() {
        if (!isInitialized()) {
            // Initialize tts.
            mTts = getTts();
        }
        // Process active notification.
        processActiveStatusBarNotifications();
    }

    private boolean isInitialized() {
        return mHandler != null;
    }

    private Handler getHandler() {
        if (mHandler == null) {
            mHandler = new Handler(Looper.getMainLooper());
        }
        return mHandler;
    }

    public void pause() {
        cancelPlayMedia();
        clearTts();
        clearAudioManager();
    }

    private void clearTts() {
        if (mTts != null) {
            mTts.stop();
            mTts.shutdown();
            mTts = null;
        }
    }

    @Override
    public void onNotificationPosted(StatusBarNotification statusBarNotification) {
        Log.i(TAG, "**********  onNotificationPosted");

        // Process new media info.
        processStatusBarNotification(statusBarNotification);
    }

    @Override
    public void onNotificationRemoved(StatusBarNotification statusBarNotification) {
        Log.i(TAG, "**********  onNotificationRemoved");

        if (isPackageRelevant(statusBarNotification)) {
            cancelPlayMedia();
        }
    }

    private void processActiveStatusBarNotifications() {
        StatusBarNotification[] activeNotifications = getActiveNotifications();
        if (activeNotifications != null) {
            for (StatusBarNotification statusBarNotification : getActiveNotifications()) {
                processStatusBarNotification(statusBarNotification);
            }
        }
    }

    private void processStatusBarNotification(StatusBarNotification statusBarNotification) {
        if (isPackageRelevant(statusBarNotification)) {
            final Extractor extractor = new Extractor();
            NotificationData notificationData = extractor.load(getApplicationContext(), statusBarNotification, new NotificationData());
            logNotificationData(statusBarNotification, notificationData);

            // Generate related media.
            final Media media = MediaFactory.createMedia(notificationData);
            if (media != null && !media.equals(mCurrentMedia)) {
                cancelPlayMedia();
                if (media.isRelevant()) {
                    mCurrentMedia = media;
                    playMediaAsync();
                    broadcastMedia(mCurrentMedia);
                }
            } else if (!isInitialized()) { // After service got stopped and started while same song is on.
                playMediaAsync();
            }
        }
    }

    private void logNotificationData(final StatusBarNotification statusBarNotification, final NotificationData notificationData) {
        String notificationInBrief = String.format(Locale.US, "{id: %d, time: %d, package: %s, Big title: %s, title: %s, " +
                        "summary: %s, info: %s, sub: %s, message: %s, message lines: %s}",
                statusBarNotification.getId(), statusBarNotification.getPostTime(),
                statusBarNotification.getPackageName(), notificationData.titleBigText,
                notificationData.titleText, notificationData.summaryText, notificationData.infoText,
                notificationData.subText, notificationData.messageText, TextUtils.concat(notificationData.messageTextLines));
        Log.i(TAG, notificationInBrief);
    }

    private void broadcastMedia(final Media media) {
        Intent intent = new Intent();
        intent.setAction(MainActivity.ACTION_BROADCAST_MEDIA);
        intent.putExtra(MainActivity.EXTRA_BROADCAST_MEDIA, media);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    private boolean isPackageRelevant(StatusBarNotification statusBarNotification) {
        return Arrays.toString(MediaApp.values()).contains(statusBarNotification.getPackageName());
    }

    private TextToSpeech getTts() {
        return new TextToSpeech(getApplicationContext(), new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {
                mTts.setLanguage(Locale.getDefault());
                mTts.setVoice(new Voice("en-us-x-sfg#male_2-local", Locale.US, Voice.QUALITY_VERY_HIGH, Voice.LATENCY_LOW,
                        false, null));
                /*mTts.setVoice(new Voice("en-us-x-sfg#female_2-local", Locale.US, Voice.QUALITY_VERY_HIGH, Voice.LATENCY_LOW,
                        false, null));*/
                mTts.setSpeechRate(mSettingsManager.getPlayMediaSpeed());
                mTts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                    @Override
                    public void onStart(String utteranceId) {

                    }

                    @Override
                    public void onDone(String utteranceId) {
                        // Abandon focus.
                        getAudioManager().abandonAudioFocus(audioFocusChangeListener);
                    }

                    @Override
                    public void onError(String utteranceId) {

                    }
                });
            }
        });
    }

    private void playMediaAsync() {
        getHandler().postDelayed(createPlayRunnable(), mSettingsManager.getPlayMediaDelayInMilliseconds());
    }

    private void cancelPlayMedia() {
        // Cancel previous tasks if applicable.
        if (mHandler != null) {
            mHandler.removeCallbacksAndMessages(null);
            mHandler = null;
        }
    }

    private Runnable createPlayRunnable() {
        return new Runnable() {
            @Override
            public void run() {
                playMedia();
                if (mHandler != null) {
                    mHandler.postDelayed(this, mSettingsManager.getPlayMediaIntervalInMilliseconds());
                }
            }
        };
    }

    private void playMedia() {
        if (mCurrentMedia != null && mCurrentMedia.isRelevant()) {
            // Prepare speech.
            String newSpeech = String.format("You listen to %s by %s", mCurrentMedia.getTitle(),
                    mCurrentMedia.getArtist());

            // Request audio focus for playback
            int result = getAudioManager().requestAudioFocus(audioFocusChangeListener,
                    // Use the notification stream.
                    AudioManager.STREAM_NOTIFICATION,
                    // Request audio focus.
                    AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK);

            if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED && mTts != null) {
                // Speak music notification info.
                Log.i(TAG, "Speech = " + newSpeech);
                mTts.speak(newSpeech, TextToSpeech.QUEUE_FLUSH, null, "UniqueID");
            }
        }
    }

    AudioManager.OnAudioFocusChangeListener audioFocusChangeListener = new AudioManager.OnAudioFocusChangeListener() {
        public void onAudioFocusChange(int focusChange) {
            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT) {

            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {

            } else if (focusChange == AudioManager.AUDIOFOCUS_LOSS) {

            }
        }
    };

    private AudioManager getAudioManager() {
        if (mAudioManager == null) {
            mAudioManager = (AudioManager) getApplicationContext().getSystemService(Context.AUDIO_SERVICE);
        }
        return mAudioManager;
    }

    private void clearAudioManager() {
        if (mAudioManager != null) {
            mAudioManager.abandonAudioFocus(audioFocusChangeListener);
            mAudioManager = null;
            audioFocusChangeListener = null;
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (isInitialized()) {
            if (key.equals(SettingsManager.PLAY_MEDIA_DELAY) || key.equals(SettingsManager.PLAY_MEDIA_INTERVAL)) {
                cancelPlayMedia();
                playMediaAsync();
            } else if (key.equals(SettingsManager.PLAY_MEDIA_SPEED) && mTts != null) {
                mTts.setSpeechRate(mSettingsManager.getPlayMediaSpeed());
            }
        }
    }

    public class LocalBinder extends Binder {
        public MediaNotificationListenerService getServiceInstance() {
            return MediaNotificationListenerService.this;
        }
    }
}

File: app/src/main/java/com/botob/ulisten2/services/MediaNotificationListenerService.kt
package com.botob.ulisten2.services

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener
import android.os.*
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.speech.tts.Voice
import android.text.TextUtils
import android.util.Log
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.botob.ulisten2.MainActivity
import com.botob.ulisten2.R
import com.botob.ulisten2.media.*
import com.botob.ulisten2.notification.Extractor
import com.botob.ulisten2.notification.NotificationData
import com.botob.ulisten2.preferences.SettingsManager
import java.util.*


class MediaNotificationListenerService : NotificationListenerService(),
    OnSharedPreferenceChangeListener {
    companion object {
        private val TAG = MediaNotificationListenerService::class.java.simpleName

        private const val NOTIFICATION_ACTION_PLAY = "com.botob.ulisten2.action.play"
        private const val NOTIFICATION_ACTION_PAUSE = "com.botob.ulisten2.action.pause"
    }

    private var currentMedia: Media? = null

    private lateinit var settingsManager: SettingsManager

    private val handler: Handler by lazy {
        Handler(Looper.getMainLooper())
    }

    private lateinit var tts: TextToSpeech

    var audioFocusChangeListener: OnAudioFocusChangeListener? =
        OnAudioFocusChangeListener { focusChange ->
            when (focusChange) {
                AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                }
                AudioManager.AUDIOFOCUS_GAIN -> {
                }
                AudioManager.AUDIOFOCUS_LOSS -> {
                }
            }
        }
    private lateinit var audioManager: AudioManager

    override fun onCreate() {
        Log.i(TAG, "onCreate")
        super.onCreate()

        // Get the audio manager.
        audioManager = getSystemService(AUDIO_SERVICE) as AudioManager

        // Instantiate the settings manager.
        settingsManager = SettingsManager(this)

        // Register this service to listen to preference changes.
        settingsManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)

        tts = TextToSpeech(this) {
            if (it != TextToSpeech.SUCCESS) {
                Log.e(TAG, "tts: initialization failed with status $it")
            }

            tts.language = Locale.getDefault()
            tts.voice = Voice(
                "en-us-x-sfg#male_2-local", Locale.US, Voice.QUALITY_VERY_HIGH, Voice.LATENCY_LOW,
                false, null
            )
            /*tts.voice = new Voice("en-us-x-sfg#female_2-local", Locale.US, Voice.QUALITY_VERY_HIGH, Voice.LATENCY_LOW,
                    false, null);*/
            tts.setSpeechRate(settingsManager.playMediaSpeed)
            tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onStart(utteranceId: String) {}
                override fun onDone(utteranceId: String) {
                    // Abandon focus.
                    audioManager.abandonAudioFocus(audioFocusChangeListener)
                }

                override fun onError(utteranceId: String) {}
            })
        }
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Log.i(TAG, "onStartCommand")

        when (intent.action) {
            NOTIFICATION_ACTION_PLAY -> {
                settingsManager.playServiceEnabled = true
            }
            NOTIFICATION_ACTION_PAUSE -> {
                settingsManager.playServiceEnabled = false
            }
        }

        return START_STICKY
    }

    override fun onDestroy() {
        Log.i(TAG, "onDestroy")
        settingsManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)

        clearTts()
        clearAudioManager()

        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? {
        Log.i(TAG, "onBind")
        resume()
        return if (SERVICE_INTERFACE == intent.action) {
            super.onBind(intent)
        } else {
            LocalBinder()
        }
    }

    override fun onRebind(intent: Intent) {
        super.onRebind(intent)
        Log.i(TAG, "onRebind")
    }

    override fun onUnbind(intent: Intent): Boolean {
        Log.i(TAG, "onUnbind")
        return true
    }

    fun resume() {
        processActiveStatusBarNotifications()
    }

    fun pause() {
        cancelPlayMedia()
        displayForegroundNotification() // Update notification by removing title.

        stopForeground(false)
    }

    private fun clearTts() {
        tts.stop()
        tts.shutdown()
    }

    override fun onNotificationPosted(statusBarNotification: StatusBarNotification) {
        Log.i(TAG, "**********  onNotificationPosted")

        if (settingsManager.playServiceEnabled) {
            // Process new media info.
            processStatusBarNotification(statusBarNotification)
        }
    }

    override fun onNotificationRemoved(statusBarNotification: StatusBarNotification) {
        Log.i(TAG, "**********  onNotificationRemoved")
        if (isPackageRelevant(statusBarNotification)) {
            cancelPlayMedia()
        }
    }

    private fun displayForegroundNotification() {
        val channelId = createNotificationChannel()

        val notificationIntent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(this, 0, notificationIntent, 0)

        val views = RemoteViews(packageName, R.layout.status_bar)

        if (settingsManager.playServiceEnabled) {
            views.setImageViewResource(R.id.play_media_action, R.drawable.ic_pause_white_24dp)
            views.setOnClickPendingIntent(
                R.id.play_media_action,
                getPendingIntent(NOTIFICATION_ACTION_PAUSE)
            )

            views.setTextViewText(R.id.status_bar_track_name, currentMedia?.title)
            views.setTextViewText(R.id.status_bar_artist_name, currentMedia?.artist)
        } else {
            views.setImageViewResource(R.id.play_media_action, R.drawable.ic_play_white_24dp)
            views.setOnClickPendingIntent(
                R.id.play_media_action,
                getPendingIntent(NOTIFICATION_ACTION_PLAY)
            )

            views.setTextViewText(
                R.id.status_bar_track_name,
                getString(R.string.notification_title_paused)
            )
            views.setTextViewText(R.id.status_bar_artist_name, null)
        }

        val title = if (settingsManager.playServiceEnabled) {
            getString(R.string.notification_title_playing)
        } else {
            getString(R.string.notification_title_paused)
        }

        val notification = NotificationCompat.Builder(this, channelId)
            .setSmallIcon(R.mipmap.ic_launcher)
            .setContent(views)
            .setContentIntent(pendingIntent)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setWhen(0)
            .build()
        startForeground(1001, notification)
    }

    /**
     * Creates a notification channel for a foreground service.
     */
    private fun createNotificationChannel(): String {
        val channelId = "com.botob.ulisten2.channel.playmedia"
        val channelName = "UListen2 Play Service"
        NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_NONE).also {
            it.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            val notificationManager =
                getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(it)
        }
        return channelId
    }

    private fun getPendingIntent(action: String): PendingIntent {
        val serviceIntent = Intent(this, MediaNotificationListenerService::class.java).also {
            it.action = action
        }
        return PendingIntent.getService(this, 0, serviceIntent, PendingIntent.FLAG_UPDATE_CURRENT)
    }

    private fun processActiveStatusBarNotifications() {
        val activeNotifications = activeNotifications
        if (activeNotifications != null) {
            for (statusBarNotification in getActiveNotifications()) {
                processStatusBarNotification(statusBarNotification)
            }
        }
    }

    private fun processStatusBarNotification(statusBarNotification: StatusBarNotification) {
        if (isPackageRelevant(statusBarNotification)) {
            val extractor = Extractor()
            val notificationData =
                extractor.load(applicationContext, statusBarNotification, NotificationData())
            logNotificationData(statusBarNotification, notificationData)

            // Generate related media.
            val media = MediaFactory.createMedia(notificationData)
            if (media != null && media != currentMedia) {
                cancelPlayMedia()
                if (media.isRelevant) {
                    currentMedia = media
                    playMediaAsync()
                    currentMedia?.let { broadcastMedia(it) }
                }
            }
        }
        displayForegroundNotification()
    }

    private fun logNotificationData(
        statusBarNotification: StatusBarNotification,
        notificationData: NotificationData?
    ) {
        val notificationInBrief = String.format(
            Locale.US,
            "{id: %d, time: %d, package: %s, Big title: %s, title: %s, " +
                    "summary: %s, info: %s, sub: %s, message: %s, message lines: %s}",
            statusBarNotification.id,
            statusBarNotification.postTime,
            statusBarNotification.packageName,
            notificationData!!.titleBigText,
            notificationData.titleText,
            notificationData.summaryText,
            notificationData.infoText,
            notificationData.subText,
            notificationData.messageText,
            TextUtils.concat(*notificationData.messageTextLines)
        )
        Log.i(TAG, notificationInBrief)
    }

    private fun broadcastMedia(media: Media) {
        val intent = Intent()
        intent.action = MainActivity.ACTION_BROADCAST_MEDIA
        intent.putExtra(MainActivity.EXTRA_BROADCAST_MEDIA, media)
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent)
    }

    private fun isPackageRelevant(statusBarNotification: StatusBarNotification): Boolean {
        return MediaApp.values().contentToString().contains(statusBarNotification.packageName)
    }

    private fun playMediaAsync() {
        handler.postDelayed(
            createPlayRunnable(),
            settingsManager.playMediaDelayInMilliseconds.toLong()
        )
    }

    private fun cancelPlayMedia() {
        // Cancel previous tasks if applicable.
        handler.removeCallbacksAndMessages(null)
        currentMedia = null
    }

    private fun createPlayRunnable(): Runnable {
        return object : Runnable {
            override fun run() {
                playMedia()
                handler.postDelayed(this, settingsManager.playMediaIntervalInMilliseconds.toLong())
            }
        }
    }

    private fun playMedia() {
        currentMedia?.run {
            if (isRelevant) {
                // Prepare speech.
                val newSpeech = String.format("You listen to %s by %s", title, artist)

                // Request audio focus for playback
                val result = audioManager.requestAudioFocus(
                    audioFocusChangeListener,  // Use the notification stream.
                    AudioManager.STREAM_NOTIFICATION,  // Request audio focus.
                    AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK
                )
                if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                    // Speak music notification info.
                    Log.i(TAG, "Speech = $newSpeech")
                    tts.speak(newSpeech, TextToSpeech.QUEUE_FLUSH, null, "UniqueID")
                }
            }
        }
    }

    private fun clearAudioManager() {
        audioManager.abandonAudioFocus(audioFocusChangeListener)
        audioFocusChangeListener = null
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (key == SettingsManager.PLAY_MEDIA_DELAY || key == SettingsManager.PLAY_MEDIA_INTERVAL) {
            cancelPlayMedia()
            playMediaAsync()
        } else if (key == SettingsManager.PLAY_MEDIA_SPEED) {
            tts.setSpeechRate(settingsManager.playMediaSpeed)
        } else if (key == SettingsManager.PLAY_SERVICE_ENABLED) {
            if (settingsManager.playServiceEnabled) {
                resume()
            } else {
                pause()
            }
        }
    }

    inner class LocalBinder : Binder() {
        val serviceInstance: MediaNotificationListenerService
            get() = this@MediaNotificationListenerService
    }
}